### PR TITLE
44561: New tables and ETL updates for viruses

### PR DESCRIPTION
**Pull request by labkey-klum**
#### Rationale
Dataspace has requested the addition of three new tables to track virus data/metadata. For now the tables will be exposed in LKS and populated through the importArchive and loadApplication ETLs with TSV data. This change does not integrate the new data into the CDS application at this time but we may do so later.

Schema design has been tracked in this ticket:

https://www.labkey.org/Dataspace/Support%20Tickets/issues-details.view?issueId=44561&_docid=issue%3A44561

#### Changes
- Upgrade scripts to create the new tables, including their import_ counterparts
- Update ETLs to import TSV data and load into the CDS schema tables
- Test .tsv data for the new tables

### CODE REVIEW

File: src/org/labkey/cds/CDSModule.java [hunk 1]
```diff
@@ -183,7 +183,7 @@ public String getName()
     @Override
     public @Nullable Double getSchemaVersion()
     {
-        return 21.003;
+        return 21.004;
```
Comment (labkey-adam) [inline src/org/labkey/cds/CDSModule.java]: Needs to be 22.001

**Review by labkey-adam — CHANGES_REQUESTED** [review]
Since we've bumped all modules to 22.000 on develop, this SQL script needs to be `cds-22.000-22.001.sql`, otherwise it won't run when it's merged to develop.

File: resources/schemas/cds.xml [hunk 2]
```diff
@@ -1412,4 +1412,84 @@
...
+            <column columnName="virus_type"/>
+            <column columnName="virus_species"/>
+            <column columnName="clade"/>
+            <column columnName="neutralization_tier"/>
+            <column columnName="container"/>
+        </columns>
+    </table>
+    <table tableName="import_virus_lab_id" tableDbType="TABLE">
+        <columns>
+            <column columnName="rowId"/>
```
Comment (XingY) [inline resources/schemas/cds.xml]: cds typically don't use rowId as far as I remember, is there a specific reason to use one here?

File: resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    virus_host_cell VARCHAR(250),
+    virus_plot_label VARCHAR(250),
+    virus_type VARCHAR(250),
+    virus_species VARCHAR(250),
+    clade VARCHAR(250),
+    neutralization_tier VARCHAR(250),
+    container ENTITYID NOT NULL,
+
+    CONSTRAINT PK_ImportVirusMetadataAll PRIMARY KEY (cds_virus_id),
+    CONSTRAINT UQ_ImportVirusMetadataAll UNIQUE (virus_full_name)
```
Comment (XingY) [inline resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql]: Should the unique constraint include container also?

File: resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    rowId SERIAL,
+    cds_virus_id INTEGER NOT NULL,
+    lab_code VARCHAR(250),
+    lab_virus_id INTEGER,
+    lab_virus_id_variable_name VARCHAR(250),
+    harvest_date TIMESTAMP,
+    container ENTITYID NOT NULL,
+
+    CONSTRAINT PK_VirusLabId PRIMARY KEY (rowId),
+    CONSTRAINT FK_PK_VirusLabId FOREIGN KEY (cds_virus_id) REFERENCES cds.virus_metadata_all(cds_virus_id)
```
Comment (XingY) [inline resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql]: FK should container "container" also

File: resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+(
+    rowId SERIAL,
+    cds_virus_id INTEGER NOT NULL,
+    lab_code VARCHAR(250),
+    lab_virus_id INTEGER,
+    lab_virus_id_variable_name VARCHAR(250),
+    harvest_date TIMESTAMP,
+    container ENTITYID NOT NULL,
+
+    CONSTRAINT PK_VirusLabId PRIMARY KEY (rowId),
```
Comment (XingY) [inline resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql]: The consistent way to define PK for metadata tables in CDS would be (cds_virus_id, container), unless there is a need for rowId that I don't know about.

**Review by XingY — CHANGES_REQUESTED** [review]

File: resources/schemas/cds.xml (hunk 2 above)
Comment (labkey-klum) [reply resources/schemas/cds.xml]: that's true, I had added that before Jason gave me the final schema & FK's

File: resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql (hunk 3 above)
Comment (labkey-klum) [reply resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql]: yup

File: resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql (hunk 4 above)
Comment (labkey-klum) [reply resources/schemas/dbscripts/postgresql/cds-21.003-21.004.sql]: yup

**Review by labkey-adam — APPROVED** [review]
Script name and schema version look good

File: test/sampledata/dataspace/etlLogs/CDS_CDSImport_2022-01-13_17-26-48.etl.log [hunk 6]
```diff
@@ -0,0 +1,104 @@
+13 Jan 2022 17:26:50,016 INFO : Starting to run task 'CDS:task:{CDS}/CDSImport#CDSImport' at location 'webserver'
```
Comment (XingY) [inline test/sampledata/dataspace/etlLogs/CDS_CDSImport_2022-01-13_17-26-48.etl.log]: those etl logs files should not be checked in

**Review by XingY — COMMENTED** [review]
please delete the etl logs

File: test/sampledata/dataspace/etlLogs/CDS_CDSImport_2022-01-13_17-26-48.etl.log (hunk 6 above)
Comment (labkey-klum) [reply test/sampledata/dataspace/etlLogs/CDS_CDSImport_2022-01-13_17-26-48.etl.log]: oops, my bad

**Review by XingY — APPROVED** [review]

**Pull request merged**